Source/MLX/MLXArray.swift

```diff
@@ -163,12 +206,59 @@ public final class MLXArray {
     public func item<T: HasDType>(_ type: T.Type) -> T {
         precondition(self.size == 1)
 
+        // special cases for reading integers and floats from (roughly)
+        // same typed arrays -- this avoids doing a conversion which
+        // might end up as an unexpected operation that would mess up
+        // async evaluation
```

Comment:
Related to the other type conversion (seen while working on LLMs), the python code does this:

```python
    while True:
        next_y, next_logprobs = _step(y)
        mx.async_eval(next_y)
        yield y.item(), logprobs
        y, logprobs = next_y, next_logprobs
```

The equivalent swift code in the same structure is:

```swift
    mutating public func next() -> Int? {
        let nextY = step(previous: y)
         asyncEval([nextY])
        
        let previousY = y
        y = nextY
        return previousY.item(Int.self)
    }
```

that is a little unwieldy (because of the lack of yield), so this is equivalent:

```swift
    mutating public func next() -> Int? {
        // save current value -- this will be returned
        let previousY = y

        // compute the next state and async eval the next token
        y = step(previous: previousY)
        asyncEval(y)

        return previousY.item(Int.self)
    }
```

So the problem here is that we do an asynchronous evaluation on `y` and then return `previousY.item()`.  Except that `item()` may end up doing a type conversion, which will have to enqueue another operation that will run _after_ the `asyncEval()`.

In this case the dtype of `previousY` is actually `uint32` but we need it as an `Int` for the tokenizer.

This code will let us do trivial type conversion between related int/float types without needing a type conversion operation.  This is closer to the python code which doesn't care what the dtype is -- it simply returns a value with the appropriate type.

TLDR: this helps us avoid traps when doing `asyncEval(next); return previous.item()`

Reply:
Indeed that is much more efficient than doing a scalar conversion with an asType.